### PR TITLE
Rework go-ipfs to use content IDs version 0

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is a refactor to replace all object identifiers code with content ID version 0. This is a drop in replacement for the existing key.Key and will not change any outward behaviour. This is just a preparatory refactor for the switchover to CIDv1 and ipld

### CODE REVIEW

**Comment by whyrusleeping** [comment]
Note: this is not quite complete yet

File: blocks/blockstore/bloom_cache.go [hunk 1]
```diff
@@ -132,10 +132,11 @@ func (b *bloomcache) PutMany(bs []blocks.Block) error {
 			good = append(good, block)
 		}
 	}
-	err := b.blockstore.PutMany(bs)
+
+	err := b.blockstore.PutMany(good)
 	if err == nil {
-		for _, block := range bs {
-			b.bloom.AddTS([]byte(block.Key()))
+		for _, blk := range good {
```
Comment (Kubuxu) [inline blocks/blockstore/bloom_cache.go]: This is resolved in https://github.com/ipfs/go-ipfs/pull/3162

File: blocks/blockstore/arc_cache.go [hunk 2]
```diff
@@ -99,12 +99,14 @@ func (b *arccache) PutMany(bs []blocks.Block) error {
 			good = append(good, block)
 		}
 	}
-	err := b.blockstore.PutMany(bs)
+
+	err := b.blockstore.PutMany(good)
```
Comment (Kubuxu) [inline blocks/blockstore/arc_cache.go]: This too.

**Comment by Kubuxu** [comment]
I went over it, can't find anything wrong with it at first sight but I would love if @lgierth went over it too.


(also it is in conflict with master now, probably due to changes in the blockstore code I pointed out).

**Comment by whyrusleeping** [comment]
One change i could probably do separately is the `Object` abstraction in `Blockservice`. It doesnt *need* to be done here, but it does simplify things a bit I think. If anyone thinks that it should be done separately I will go ahead and do that.

Other than that, I think that the pin code changes should be looked over closely. @lgierth if you could pay special attention there that would be great :)

**Comment by lgierth** [comment]
> This is a refactor to replace all object identifiers code with content ID version 0.

Is the following accurate? It doesn't replace *all* usage of key.Key -- from the block layer on inwards we assume raw data (where s/Data/RawData). The layer above that will be the one dealing with the different semantics of various CIDs (soon).

I'm looking over `git diff pin/` with this assumption in mind, mh, tricky.

---

```go
// an Object is simply a typed block
type Object interface {
	Cid() *cid.Cid
	blocks.Block
}
```

:+1:

**Comment by whyrusleeping** [comment]
@lgierth yeah, youre right. the `key.Key` is still around at the block layer. We're talking about making the CIDs go all the way down to disk, it has its pros and cons, but it would drastically increase the complexity of this change, and is definitely something we can do in a follow up PR.

File: pin/pin.go [hunk 3]
```diff
@@ -259,90 +269,99 @@ func (p *pinner) isPinnedWithType(k key.Key, mode PinMode) (string, bool, error)
...
-	checkChildren = func(rk key.Key, parentKey key.Key) error {
+	var checkChildren func(*cid.Cid, *cid.Cid) error
+	checkChildren = func(rk, parentKey *cid.Cid) error {
 		parent, err := p.dserv.Get(context.Background(), parentKey)
 		if err != nil {
 			return err
 		}
 		for _, lnk := range parent.Links {
-			k := key.Key(lnk.Hash)
+			c := cid.NewCidV0(lnk.Hash)
```
Comment (lgierth) [inline pin/pin.go]: I assume this is explicitly v0 because we don't wanna start using v1 right now already

File: pin/pin.go [hunk 4]
```diff
@@ -259,90 +269,99 @@ func (p *pinner) isPinnedWithType(k key.Key, mode PinMode) (string, bool, error)
...
 	}
 }
 
+func cidSetWithValues(cids []*cid.Cid) *cid.Set {
+	out := cid.NewSet()
+	for _, c := range cids {
+		out.Add(c)
+	}
+	return out
+}
```
Comment (lgierth) [inline pin/pin.go]: Is this a candidate for moving to go-cid?

**Comment by lgierth** [comment]
LGTM :+1: 

- `pin/` could see the functions named after Key (e.g. `Pinner.DirectKeys()`) renamed
- some of `pin/` still uses Key, which is mostly the GC code -- I can understand why you didn't wanna touch it *right now*, it can get messy and we can take care of it later.
- these two made it a bit confusing at first, but I think it's good to go

File: pin/pin.go (hunk 3 above)
Comment (whyrusleeping) [inline pin/pin.go]: Yeah, but thinking about it, we could change that to just be a `cid.Cast` and it would work just fine

File: pin/pin.go (hunk 4 above)
Comment (whyrusleeping) [inline pin/pin.go]: Ah, yeah. thats a great point

File: pin/pin.go (hunk 3 above)
Comment (lgierth) [inline pin/pin.go]: Nah it's fine, one silly mistake and we'll have to maintain it :D:D

**Comment by lgierth** [comment]
And I guess this PR has the nice side-effect of reducing the overall dependency on the blocks package :)

File: pin/pin.go (hunk 4 above)
Comment (jbenet) [inline pin/pin.go]: did this get moved? dont see commits after

**Pull request merged**